lib/compat.php

```diff
@@ -304,3 +304,235 @@ function gutenberg_warn_classic_about_blocks() {
 		</script>
 	<?php
 }
+
+/**
+ * Add 'actual_size' to the prepared attachment data for the Media Library.
+ *
+ * Needed as `wp_prepare_attachment_for_js()` (for the media modal) constrains the image sizes
+ * to the theme's `$content_width`.
+ *
+ * @param array   $response   Array of prepared attachment data.
+ * @param WP_Post $attachment Attachment object.
+ * @param array   $meta       Array of attachment meta data.
+ * @return array Array of prepared attachment data.
+ */
+function gutenberg_prepare_attachment_for_js( $response, $attachment, $meta ) {
+	if ( ! empty( $response['type'] ) && $response['type'] === 'image' && ! empty( $response['sizes'] ) ) {
+		foreach ( $response['sizes'] as $size_name => $size ) {
+			if ( array_key_exists( $size_name, $meta['sizes'] ) ) {
+				$response['sizes'][ $size_name ]['actual_size'] = array(
+					'width' => (int) $meta['sizes'][ $size_name ]['width'],
+					'height' => (int) $meta['sizes'][ $size_name ]['height'],
+				);
+			}
+		}
+	}
+
+	return $response;
+}
+add_filter( 'wp_prepare_attachment_for_js', 'gutenberg_prepare_attachment_for_js', 10, 3 );
+
+/**
+ * Warm the object cache with post and meta information for all found
+ * image blocks to avoid making individual database calls
+ * (similarly to `wp_make_content_images_responsive()`).
+ *
+ * @access private
+ * @since 4.x.x
+ *
+ * @param string $content The post content.
+ * @return string $content Unchanged post content.
+ */
+function _gutenberg_cache_images_meta( $content ) {
+	// Need to find all image blocks and get the attachment IDs from them BEFORE the parser is run
+	// so we can get the image attachments meta all at once from the DB.
+	if ( preg_match_all( '/^<!-- wp:image {.*"id":(\d+),.*} -->$/m', $content, $matches ) ) {
+		_prime_post_caches( $matches[1], false, true );
+	}
+
+	return $content;
+}
```

Comment:
is there a reason for this not to be hooking into the block filter so that we can get the actual attributes?

```php
function _gutenberg_cache_images_meta( $html, $block ) {
	if ( 'core/image' === $block[ 'blockName' ] ) {
		_prime_post_caches( $block[ 'attrs' ][ 'id' ], /* why aren't these */ false, /* commented? */ true );
	}

	return $html;
}

add_filter( 'rendered_block', '_gutenberg_cache_images_meta', 3 );
```

Reply:
The purpose here is to get the IDs of all images (where we will do srcset and sizes) from the post before we parse the blocks. That's needed to get all image attachment meta in one go from the DB as that's faster. 

So we need an array with all attachment IDs for the images used in the post. This is equivalent to https://core.trac.wordpress.org/browser/tags/4.9.8/src/wp-includes/media.php#L1336.

That should also include images from "Cover" and "Media and Text" blocks, and probably the Gallery block (if images there are "hardcoded"). Any ideas how to do that without a regex are very welcome :)

Reply:
Previously: https://github.com/WordPress/gutenberg/pull/11377#discussion_r234008623

I'm not sure I understand why this would need to occur prior to the parse. If it at least occurs prior to _any block evaluating its `render_callback`_, wouldn't that be equally sufficient to warm the cache?

Reply:
>If it at least occurs prior to _any block evaluating its `render_callback`_, wouldn't that be equally sufficient to warm the cache?

Since this doesn't exist as an option today anyways (requires filter on [this value](https://github.com/WordPress/WordPress/blob/bf1668333f2da64036db1b10604de18bbbfb8099/wp-includes/blocks.php#L247)), we probably can't do anything of this sort right now.

Should follow-up with a Trac ticket as appropriate.

Reply:
Yeah, this doesn't do anything if we do it "per block". It makes sense only when it happens for all of post_content at once.

Reply:
if we hook into the block-level filter (the design of which I favor in #10108) then we can track the parse through the blocks and then prime the cache on the `do_blocks` filter which occurs after parsing and rendering is complete.

that is, we could imagine something like this…

```php

function _gutenberg_cache_images_meta( $html, $block ) {
	if ( $block['blockName'] === 'core/image' ) {
		$this->add_image( $block['attrs']['id'] );
	}

	return $html;
}
add_filter( 'rendered_block', '_gutenberg_cache_images_meta', 3 );

function _gutenberg_prime_cache() {
	if ( empty( $this->images ) ) {
		return;
	}

	// whatever command does all images in one go
	_prime_post_caches( $this->images, false, true, true, true, false, false );
}
add_filter( 'do_blocks', '_gutenberg_prime_cache' );
```

and this because we know that the final `do_blocks` filter runs _after_ block-level processing.

Reply:
noting that this is still based on a RegExp instead of using the per-block and per-document filters

Reply:
Another option would be to "unify" this with the "old" cache priming. We have the (required) `wp-image-####` className on all images, including in Cover and Media and Text blocks.

Will need to abstract [`wp_make_content_images_responsive()`](https://core.trac.wordpress.org/browser/tags/4.9.8/src/wp-includes/media.php#L1305) a bit, move the caching out and run it earlier on `the_content` filter.

On the other hand `_prime_post_caches()` is "clever enough" to not fetch posts and meta that are already cached so running two pre-caching methods won't "hurt" anything.